add yearn vaults v3 docs (#291)
yearn/yearn-devdocs: docs/getting-started/products/yvaults/v3.md

- `deposit(assets, receiver)`: Deposits `assets` amount of tokens into the strategy. It can be restricted to the vault only or be open.
- `maxWithdraw(address)`: View how many assets can be taken from the vault at any given point in time.
- `withdraw(assets, receiver, owner)`: Withdraws `assets` amount of tokens from the strategy.
- `balanceOf(address)`: View the number of shares of the strategy that the address has.

This means that the vault can deposit into any ERC4626 vault and that a non-compliant strategy can be implemented provided that these functions have been implemented (even if they are in a non-ERC4626 compliant way). 

Anything else is left to the strategy writer. However, to make security review easier, Yearn's template has the following optional functions: 
- `tend()`: a function that will be called by bots. It will do anything to maintain a position, such as act on certain triggers.
- `tendTrigger()`: implementation which triggers bots that will call tend function on the contract.
- `invest()`: deposit funds into the underlying protocol.
- `emergencyFreeFunds()`: close the position and return funds to the strategy. Losses can be accepted here.

> ## ERC4626 Compliance

Reviewer: This is referenced above in a hackmd I recommend moving above or deleting that section above.

Author: the hackmd was deleted and instead it points to this section now

When deploying a new vault, it requires the following parameters:
- **asset:** Address of the ERC20 token that can be deposited in the vault.
- **name:** Name of Shares as described in the ERC20.
- **symbol:** Symbol of Shares in the ERC20.
- **role_manager:** Account that can assign and revoke Roles.
- **PROFIT_MAX_UNLOCK_TIME:** Max amount of time profit will be distributed over.

## Normal Operation

### Deposits / Mints
Users can deposit ASSET tokens to receive yvTokens (SHARES).

Deposits are limited under `depositLimit` and shutdown parameters. Read below for details.

### Withdrawals / Redeems
Users can redeem their shares at any time if there is liquidity available. 

The redeem function will check if there are enough idle funds to serve the request. If there are not enough, it will revert. 

Optionally, a user can specify a list of strategies to withdraw from. If a list of strategies is passed, the vault will try to withdraw from them.

If not enough funds have been recovered to honor the full request, the transaction will revert.

### Vault Shares
Vault shares are ERC20 transferable yield-bearing tokens.

> They are ERC4626 compliant. Please read [ERC4626 compliance](https://hackmd.io/cOFvpyR-SxWArfthhLJb5g#ERC4626-compliance) to understand the implications. 

Author: fixed!

Strategy migration is the process of replacing an existing strategy with a new one, which will inherit all of the old strategy's parameters, including its debt.

### Debt Management
Callers take this responsibility with the `DEBT_MANAGER` role.

#### Setting minimum idle funds
The `DEBT_MANAGER` can specify how much funds the vault should try to have reserved to serve withdrawal requests.

These funds will remain in the vault unless requested by a Depositor.

#### Setting maximum debt for a specific strategy
The maximum amount of tokens the vault will allow a strategy to owe at any moment.

Stored in `strategies[strategy].max_debt`.

When a debt rebalance is triggered, the Vault will cap the new target debt to this number (max_debt).

#### Rebalance Debt
The vault sends and receives funds to/from strategies. The function updateDebt(strategy, target_debt) will set the current_debt of the strategy to target_debt (if possible).

If the strategy currently has less debt than the target_debt, the vault will send funds to it.

The vault checks that the `minimumTotalIdle` parameter is respected (i.e. there's at least a certain amount of funds in the vault).

> The vault will request back funds if the strategy has more debt than the `max_debt`. These funds may be locked in the strategy, which will result in the strategy returning fewer funds than requested by the vault. 

Reviewer: `These funds may be locked in the strategy` and if they are not locked in the strategy?

## Roles
Vault functions that are permissioned will be callable by accounts with specific roles. 

These are: 
- `STRATEGY_MANAGER:` role for accounts that can add, remove or migrate strategies.
- `DEBT_MANAGER:` role for accounts that can rebalance and manage debt related params.
- `EMERGENCY_MANAGER:` role for accounts that can activate the shutdown mode.
- `ACCOUNTING_MANAGER:` role for accounts that can process reports (save strategy p&l) and change accountant parameters.

Every role can be filled by an EOA, multisig, or other smart contracts. Each role can be filled by several accounts.

> The account that manages roles is a single account, set in `role_manager`.

Reviewer: why is `role_manager` not capped while `STRATEGY_MANAGER` is capped?

Reviewer: I see that makes sense not sure if it needs to be added in this document though

Reviewer: Yes that makes sense thanks, but I dont think this should be added in this document

> This `role_manager` can be an externally owned account, a multi-signature wallet, or a Governance Module that relays calls. 

Reviewer: why is governance module capped?

## Strategy Minimum API
Strategies are completely independent smart contracts that can be implemented following the proposed template or in any other way.

In any case, to be compatible with the vault, they need to implement the following functions, which are a subset of ERC4626 vaults: 
- `asset()`: View returning underlying asset.
- `vault()`: View returning vault this strategy is plugged to.
- `totalAssets()`: View returning current amount of assets. It can include rewards valued in `asset`.
- `maxDeposit(address)`: View returning the amount max that the strategy can take safely.
- `deposit(assets, receiver)`: Deposits `assets` amount of tokens into the strategy. It can be restricted to the vault only or be open.

> - `maxWithdraw(address)`: View returning how many assets can the vault take from the vault at any given point in time.

Reviewer: This isnt at any given point in time but just `now` time correct?

- `withdraw(assets, receiver, owner)`: Withdraws `assets` amount of tokens from the strategy.
- `balanceOf(address)`: Return the number of shares of the strategy that the address has.

This means that the vault can deposit into any ERC4626 vault and that a non-compliant strategy can be implemented provided that these functions have been implemented (even in a non ERC4626 compliant way). 

Anything else is left to the strategy writer. However, to make security review easier, the Yearn's template has the following optional functions: 

> - `tend()`: a function that will be called by bots. It will do anything to maintain a position, such as act on certain triggers.

Reviewer: `It will do anything to maintain a position` anything sounds like an odd word choice here, what are we really saying? The bots can tend anyway they can think of or?

Author: updated to use jmonteer's wording

> - `tendTrigger()`: implementation that trigger bots that will call tend function on the contract.

Reviewer: This sounds odd maybe rewrite for easier reading. Im not fully sure what tendTrigger does. This call tiggers bots, or sets up a schedule which tiggers bots?

- `invest()`: deposit funds into underlying protocol.
- `emergencyFreeFunds()`: close the position and return funds to the strategy. Losses might be accepted here.

## ERC4626 compliance

> Vault Shares are ERC4626 compliant. 

Reviewer: why is `Vault Shares` capped?

Author: fixed!

> The most important implication is that `withdraw` and `redeem` functions as presented in ERC4626, the liquidity to redeem shares will be the one in the vault. No strategies will be passed to the redeem function to withdraw from with the ERC4626 compliant `withdraw` and `redeem` function. 

Reviewer: The most important implication is that function 1 and 2 what? are in one vault?

## Emergency Operation

### Shutdown mode

> If the current roles stop fulfilling their responsibilities or something else happens, the `EMERGENCY_MANAGER` can shut down the vault.

Reviewer: `something else happens` > `an error occurs` might sound better?

Author: used jmonteers wording to update this sentence

# Vaults Version 3

Version 3 yVaults iterates on Version 2 by increasing robustness and developing yearn’s path towards further decentralization, while keeping the same proven product (yield-bearing tokens) that abstract builders and users from the hard work of yield farming. Version 3 will initially have the same functionality as Version 2, but the vaults are more robust and modular.

Version 3 yVaults improves on Version 2 by:
- Increasing vault modularity, allowing for smaller and safer pieces of code.
- Simplify strategy creation, empowering strategists and reducing the chance for errors.
- Adding new products like ySwaps which increases swap efficiency and yJuniorTranches which allows for different risk profiles.

## Smart Modules

> Smart Modules implement logic to be iterated until it can be made immutable. If any Smart Module fails, the vault can live without them just enough to return funds to depositors. **When Version 3 launches, the Smart Modules will replicate Version 2 vault behavior.**

Reviewer: `to be iterated until it can be made immutable`
Not really sure here sounds odd though.

Author: ```suggestion
Smart Modules implement core vault logic that will be iterated until they can be made immutable. If any Smart Module fails, the vault can live without them just enough to return funds to depositors. **When Version 3 launches, the Smart Modules will replicate Version 2 vault behavior.**
```

Author: @DarkGhost7 sounds better?

- **Strategy Registry:** Handles adding and removing strategies of a vault. It can also include strategy metadata.
- **Fee Manager:** Handles changing fees for vault operations.
- **HealthCheck:** Guardrail vault operations so that profit & loss reporting is always under acceptable values.

## Peripheral Modules

Peripheral Modules are a separate layer of optional functions for the vault and strategies. They are not required but they facilitate building around yVaults and creating new strategies.

- **yRouter**: Wrapper that handles deposits and withdrawals to/from the vault.
- **Yearn Lens:** Information aggregator for off-chain apps.
- **APY TWAP Oracle:** Reliable source of Yearn vaults’ past performance.
- **ySwaps:** Internal swap system currently being developed. Reduces slippage thus improving net APY.
- And any others you can come up with!

# Version 3 yVault Specification

## Definitions
- **Asset:** Any ERC20-compliant token
- **Shares:** ERC20-compliant token that tracks Asset balance in the vault for every distributor. Named yv<Asset_Symbol>
- **Depositor:** Account that holds Shares
- **Strategy:** Smart contract used to deposit in Protocols to generate yield

> - **Vault:** ERC4626 compliant smart contract that receives Assets from Depositors to distribute amongst the different Strategies added to the vault: manages accounting and Assets distribution. 

Reviewer: `manages accounting` you say manages accounting but there is an Accountant smart contract?

Author: I think @jmonteer meant that the vault handles accounting logic for all users

Author: https://github.com/yearn/yearn-devdocs/pull/291/files#diff-7fc26d081f78daefe7b3cf46c856f64bb949779358599412cd3b2250ab13bab8R38

- **Role:** Flags that an Account can do specific Vault actions. Can be fulfilled by a smart contract or an EOA.
- **Accountant:** Smart contract that receives P&L reporting and returns shares and refunds to the strategy

> # VaultV3 Specification

Reviewer: This title is posted twice at the same heading level

The Vault code has been designed as an unopinionated system to distribute depositors funds into different opportunities (aka Strategies) and robustly manage accounting.

The depositors receive shares of the different investments that can then be redeemed or used as yield-bearing tokens.

The Vault does not have a preference on any of the dimensions that should be considered when operating a vault:
- **Decentralization**: Roles can be filled by common wallets, smart contracts like multisigs or governance modules.
- **Liquidity**: Vault can have 0 liquidity or be fully liquid. It will depend on the parameters and strategies added.

> - **Security**: Vault managers can choose what strategies to add and how to do that process.

Reviewer: `managers` is this just a term or an specific role/position?

Author: good question idk

Reviewer: Ok thanks

## Example periphery contracts: 
- **Emergency module:** Receives deposits of Vault Shares and allows the contract to call the shutdown function after a certain % of total Vault Shares have been deposited.
- **Debt Allocator:** Smart contract that incentivizes APY / debt allocation optimization by rewarding the best debt allocation (see [yStarkDebtAllocator](https://github.com/jmonteer/ystarkdebtallocator)).
- **Strategy Staking Module:** Smart contract that allows players to sponsor specific strategies (so that they are added to the vault) by staking their YFI, making money if they do well and losing money if they don't.

## Deployment
We expect all the vaults available to be deployed from a Factory Contract, publicly available and callable. 

Players deploying "branded" vaults (e.g. Yearn) will use a separate registry to allow permissioned endorsement of vaults for their product

When deploying a new vault, it requires the following parameters:
- **asset:** Address of the ERC20 token that can be deposited in the vault.
- **name:** Name of Shares as described in ERC20.
- **symbol:** Symbol of Shares ERC20.
- **role_manager:** Account that can assign and revoke Roles.

> - **PROFIT_MAX_UNLOCK_TIME:** Max amount of time profit will be distributed over.

Reviewer: `PROFIT_MAX_UNLOCK_TIME` is there a reason this specific one is capped?

Optionally, a user can specify a list of strategies to withdraw from. If a list of strategies is passed, the vault will try to withdraw from them.

If not enough funds have been recovered to honor the full request, the transaction will revert.

### Vault Shares
Vault shares are ERC20 transferable yield-bearing tokens.

They are ERC4626 compliant. Please read [ERC4626 compliance](https://hackmd.io/cOFvpyR-SxWArfthhLJb5g#ERC4626-compliance) to understand the implications. 

### Accounting
The vault will evaluate profit and losses from the strategies. 

This is done by comparing the current debt of the strategy with the total assets the strategy is reporting to have. 

- If `totalAssets < currentDebt`: the vault will record a loss
- If `totalAssets > currentDebt`: the vault will record a profit

Both loss and profit will impact the strategy's debt, increasing the debt (current debt + profit) if there are profits, and decreasing its debt (current debt - loss) if there are losses.

#### Fees
The Accountant module handles fee assessment and distribution. It will report the amount of fees that need to be charged, and the vault will issue shares for that amount of fees.

#### Refunds

> Refunds are positive inflows for the vault, and are sent to the vault to reward it or compensate in anyway. The Accountant module will take care of them. 

Reviewer: `The Accountant module will take care of them.` to
`The Accountant module handles refund accounting`, maybe

The refunds are paid in the form of vault shares and will be instantly locked (and released gradually)

### Profit distribution 
Profit from different process_report calls will accumulate in a buffer. This buffer will be linearly unlocked over the locking period seconds at profit_distribution_rate. 

Profits will be locked for a max period of PROFIT_MAX_UNLOCK_TIME seconds and will be gradually distributed. To avoid spending too much gas for profit unlock, the period that profit will be locked for is a weighted average between the new profit and the previous profit.

- `new_locking_period = locked_profit * pending_time_to_unlock + new_profit * PROFIT_MAX_UNLOCK_TIME / (locked_profit + new_profit)`
- `new_profit_distribution_rate = (locked_profit + new_profit) / new_locking_period`

Losses will be offset by locked profit, if possible. That makes frontrunning losses impossible (unless loss > locked profit)

Issue of new shares due to fees will also insta-unlock profit so that price-per-share (pps) does not go down.

Both of these offsets will prevent frontrunning (as the profit was already earned and was not distributed yet)

> Instead of locking the profit, and in order to be more capital efficient (and avoid complex asset management), the vault issues shares that holds itself. Those shares are gradually burned over time (increasing pps). So instead of releasing assets, the vault will gradually burn shares.

Reviewer: what does `holds itself` mean?

- ...
```

## Definitions
- Asset: Any ERC20-compliant token
- Shares: ERC20-compliant token that tracks Asset balance in the vault for every distributor. Named yv<Asset_Symbol>
- Depositor: Account that holds Shares
- Strategy: Smart contract that is used to deposit in Protocols to generate yield
- Vault: ERC4626 compliant Smart contract that receives Assets from Depositors to then distribute them among the different Strategies added to the vault, managing accounting and Assets distribution. 
- Role: the different flags an Account can have in the Vault so that the Account can do certain specific actions. Can be fulfilled by a smart contract or an EOA.
- Accountant: smart contract that receives P&L reporting and returns shares and refunds to the strategy

## Deployment
We expect all the vaults available to be deployed from a Factory Contract, publicly available and callable. 

Players deploying "branded" vaults (e.g. Yearn) will use a separate registry to allow permissioned endorsement of vaults for their product

When deploying a new vault, it requires the following parameters:

> - asset: address of the ERC20 token that can be deposited in the vault

Author: ```suggestion
- `asset`: address of the ERC20 token that can be deposited in the vault
```

> - name: name of Shares as described in ERC20

Author: ```suggestion
- `name`: name of Shares as described in ERC20
```

> - symbol: symbol of Shares ERC20

Author: ```suggestion
- `symbol`: symbol of Shares ERC20
```

> - role_manager: account that can assign and revoke Roles

Author: ```suggestion
- `role_manager`: account that can assign and revoke Roles
```

> - profit_max_unlock_time: max amount of time profit will be locked before being distributed

Author: ```suggestion
- `profit_max_unlock_time`: max amount of time profit will be locked before being distributed
```

## Normal Operation

### Deposits / Mints

> Users can deposit ASSET tokens to receive yvTokens (SHARES).

Author: ```suggestion
Users can deposit asset tokens (ASSET) to receive yvTokens (SHARES).
```

> Deposits are limited under depositLimit and shutdown parameters. Read below for details.

Author: ```suggestion
Deposits are limited under `depositLimit` and shutdown parameters. Read below for details.
```

If totalAssets < currentDebt: the vault will record a loss
If totalAssets > currentDebt: the vault will record a profit

Both loss and profit will impact strategy's debt, increasing the debt (current debt + profit) if there are profits, decreasing its debt (current debt - loss) if there are losses.

#### Fees
Fee assessment and distribution are handled by the Accountant module. 

It will report the amount of fees that need to be charged and the vault will issue shares for that amount of fees.

There is also an optional protocol_fee that can be charged based on the configuration of the VaultFactory.vy

### Profit distribution 
Profit from different process_report calls will accumulate in a buffer. This buffer will be linearly unlocked over the locking period seconds at profit_distribution_rate. 

Profits will be locked for a max period of time of profit_max_unlock_time seconds and will be gradually distributed. To avoid spending too much gas for profit unlock, the amount of time a profit will be locked is a weighted average between the new profit and the previous profit. 

> new_locking_period = locked_profit * pending_time_to_unlock + new_profit * PROFIT_MAX_UNLOCK_TIME / (locked_profit + new_profit)

Author: ```suggestion
`new_locking_period = locked_profit * pending_time_to_unlock + new_profit * PROFIT_MAX_UNLOCK_TIME / (locked_profit + new_profit)`
```

> new_profit_distribution_rate = (locked_profit + new_profit) / new_locking_period

Author: ```suggestion
`new_profit_distribution_rate = (locked_profit + new_profit) / new_locking_period`
```

Losses will be offset by locked profit, if possible.

Issue of new shares due to fees will also unlock profit so that PPS does not go down. 

> Both of this offsets will prevent front running (as the profit was already earned and was not distributed yet)

Author: ```suggestion
Both of this offsets will prevent front running (as the profit was already earned and was not distributed yet).
```

## Vault Management

> Vault management is split into function specific roles. Each permissioned function has its own corresponding Role.

Author: ```suggestion
Vault management is split into function specific roles. Each permissioned function has its own corresponding role.
```

This means roles can be combined all to one address, each distributed to separate addresses or any combination in between

## Roles
Vault functions that are permissioned will be callable by accounts that hold specific roles. 

These are: 

> - ADD_STRATEGY_MANAGER: role than can add strategies to the vault

Author: ```suggestion
- `ADD_STRATEGY_MANAGER`: role than can add strategies to the vault.
```

> - REVOKE_STRATEGY_MANAGER: role that can remove strategies from the vault

Author: ```suggestion
- `REVOKE_STRATEGY_MANAGER`: role that can remove strategies from the vault.
```

> - FORCE_REVOKE_MANAGER: role that can force remove a strategy causing a loss

Author: ```suggestion
- `FORCE_REVOKE_MANAGER`: role that can force remove a strategy causing a loss.
```

> - ACCOUNTANT_MANAGER: role that can set the accountant that assesses fees

Author: ```suggestion
- `ACCOUNTANT_MANAGER`: role that can set the accountant that assesses fees.
```

> - QUEUE_MANAGER: role that can set the default withdrawal queue.

Author: ```suggestion
- `QUEUE_MANAGER`: role that can set the default withdrawal queue.
```

> - REPORTING_MANAGER: role that calls report for strategies

Author: ```suggestion
- `REPORTING_MANAGER`: role that calls report for strategies.
```

> - DEBT_MANAGER: role that adds and removes debt from strategies

Author: ```suggestion
- `DEBT_MANAGER`: role that adds and removes debt from strategies.
```

> - MAX_DEBT_MANAGER: role that can set the max debt for a strategy

Author: ```suggestion
- `MAX_DEBT_MANAGER`: role that can set the max debt for a strategy.
```

> - DEPOSIT_LIMIT_MANAGER: role that sets deposit limit for the vault

Author: ```suggestion
- `DEPOSIT_LIMIT_MANAGER`: role that sets deposit limit for the vault.
```

> - MINIMUM_IDLE_MANAGER: role that sets the minimum total idle the vault should keep

Author: ```suggestion
- `MINIMUM_IDLE_MANAGER`: role that sets the minimum total idle the vault should keep.
```

> - PROFIT_UNLOCK_MANAGER: role that sets the profit_max_unlock_time

Author: ```suggestion
- `PROFIT_UNLOCK_MANAGER`: role that sets the profit_max_unlock_time.
```

> - DEBT_PURCHASER # can purchase bad debt from the vault

Author: ```suggestion
- `DEBT_PURCHASER`: can purchase bad debt from the vault.
```

> - EMERGENCY_MANAGER: role that can shutdown vault in an emergency

Author: ```suggestion
- `EMERGENCY_MANAGER`: role that can shutdown vault in an emergency.
```

Every role can be filled by an EOA, multi-sig or other smart contracts. Each role can be filled by several accounts.

The account that manages roles is a single account, set in `role_manager`.

This role_manager can be an EOA, a multi-sig or a Governance Module that relays calls. 

### Strategy Management

> This responsibility is taken by callers with ADD_STRATEGY_MANAGER, REVOKE_STRATEGY_MANAGER and FORCE_REVOKE_MANAGER roles

Author: ```suggestion
This responsibility is taken by callers with `ADD_STRATEGY_MANAGER`, `REVOKE_STRATEGY_MANAGER` and `FORCE_REVOKE_MANAGER` roles
```

A vault can have strategies added, removed or forcefully removed 

Added strategies will be eligible to receive funds from the vault, when the max_debt is set to > 0

Revoked strategies will return all debt and stop being eligible to receive more. It can only be done when the strategy's current_debt is 0

Force revoking a strategy is only used in cases of a faulty strategy that cannot otherwise have its current_debt reduced to 0. Force revoking a strategy will result in a loss being reported by the vault.

#### Setting the periphery contracts

> The accountant can each be set by the ACCOUNTANT_MANAGER.

Author: ```suggestion
The accountant can each be set by the `ACCOUNTANT_MANAGER`.
```

The contract is not needed for the vault to function but are recommended for optimal use.

#### Reporting profits

> The REPORTING_MANAGER is in charge of calling process_report() for each strategy in the vault according to its own timeline

Author: ```suggestion
The `REPORTING_MANAGER` is in charge of calling process_report() for each strategy in the vault according to its own timeline
```

This call will do the necessary accounting and profit locking for the individual strategy as well as charging fees

### Debt Management
This responsibility is taken by callers with DEBT_MANAGER role

This role can increase or decrease strategies specific debt.

The vault sends and receives funds to/from strategies. The function updateDebt(strategy, target_debt) will set the current_debt of the strategy to target_debt (if possible)

If the strategy currently has less debt than the target_debt, the vault will send funds to it.

The vault checks that the `minimumTotalIdle` parameter is respected (i.e. there's at least a certain amount of funds in the vault).

If the strategy has more debt than the max_debt, the vault will request the funds back. These funds may be locked in the strategy, which will result in the strategy returning less funds than requested by the vault. 

#### Setting maximum debt for a specific strategy

> The MAX_DEBT_MANAGER can set the maximum amount of tokens the vault will allow a strategy to owe at any moment in time.

Author: ```suggestion
The `MAX_DEBT_MANAGER` can set the maximum amount of tokens the vault will allow a strategy to owe at any moment in time.
```

> Stored in strategies[strategy].max_debt

Author: ```suggestion
Stored in `strategies[strategy].max_debt`
```

When a debt re-balance is triggered, the Vault will cap the new target debt to this number (max_debt)

#### Setting the deposit limit

> The DEPOSIT_LIMIT_MANAGER is in charge of setting the deposit_limit for the vault

Author: ```suggestion
The `DEPOSIT_LIMIT_MANAGER` is in charge of setting the deposit_limit for the vault.
```

On deployment deposit_limit defaults to 0 and will need to be increased to make the vault functional

#### Setting minimum idle funds

> The MINIMUM_IDLE_MANAGER can specify how many funds the vault should try to have reserved to serve withdrawal requests

Author: ```suggestion
The `MINIMUM_IDLE_MANAGER` can specify how many funds the vault should try to have reserved to serve withdrawal requests
```

These funds will remain in the vault unless requested by a Depositor

It is recommended that if no queue_manager is set some amount of funds should remain idle to service withdrawals

#### Setting the profit unlock period

> The PROFIT_UNLOCK_MANAGER is in charge of updating and setting the profit_max_unlock_time which controls how fast profits will unlock

Author: ```suggestion
The `PROFIT_UNLOCK_MANAGER` is in charge of updating and setting the profit_max_unlock_time which controls how fast profits will unlock
```

This can be customized based on the vault based on aspects such as number of strategies, TVL, expected returns etc.

#### Setting the default queue

> The QUEUE_MANAGER has the option to set a custom default_queue if desired. The vault will arrange the default queue automatically based only on the order that strategies were added to the vault. If a different order is desired the queue manager role can set a custom queue.

Author: ```suggestion
The `QUEUE_MANAGER` has the option to set a custom default_queue if desired. The vault will arrange the default queue automatically based only on the order that strategies were added to the vault. If a different order is desired the queue manager role can set a custom queue.
```

All strategies in the default queue must have been previously added to the vault.

#### Buying Debt

> The DEBT_PURCHASER role can buy debt from the vault in return for the equal amount of `asset`.

Author: ```suggestion
The `DEBT_PURCHASER` role can buy debt from the vault in return for the equal amount of `asset`.
```

This should only ever be used in the case where governance wants to purchase a set amount of bade debt from the vault in order to not report a loss.

It still relies on convertToShares() so will only be viable if the conversion does not reflect and large negative realized loss from the strategy.

#### Shutting down the vault

> In an emergency the EMERGENCY_MANAGER can shutdown the vault

Author: ```suggestion
In an emergency the `EMERGENCY_MANAGER` can shutdown the vault
```

> This will also give the EMERGENCY_MANAGER the DEBT_MANAGER roles as well so funds can start to be returned from the strategies

Author: ```suggestion
This will also give the `EMERGENCY_MANAGER` the `DEBT_MANAGER` roles as well so funds can start to be returned from the strategies
```

## Strategy Minimum API
Strategies are completely independent smart contracts that can be implemented following the proposed template or in any other way.

In any case, to be compatible with the vault, they need to implement the following functions, which are a subset of ERC4626 vaults: 

> - asset(): view returning underlying asset

Author: ```suggestion
- `asset()`: view returning underlying asset
```

> - totalAssets(): view returning current amount of assets. It can include rewards valued in `asset` ¡

Author: ```suggestion
- `totalAssets()`: view returning current amount of assets. It can include rewards valued in `asset` ¡
```

> - maxDeposit(address): view returning the amount max that the strategy can take safely

Author: ```suggestion
- `maxDeposit(address)`: view returning the amount max that the strategy can take safely
```

> - deposit(assets, receiver): deposits `assets` amount of tokens into the strategy. it can be restricted to vault only or be open

Author: ```suggestion
- `deposit(assets, receiver)`: deposits `assets` amount of tokens into the strategy. it can be restricted to vault only or be open
```

> - maxWithdraw(address): view returning how many asset can the vault take from the vault at any given point in time

Author: ```suggestion
- `maxWithdraw(address)`: view returning how many asset can the vault take from the vault at any given point in time
```

> - withdraw(assets, receiver, owner): withdraws `assets` amount of tokens from the strategy

Author: ```suggestion
- `withdraw(assets, receiver, owner)`: withdraws `assets` amount of tokens from the strategy
```

> - redeem(shares, receiver, owner): redeems `shares` of the strategy for the underlying asset.

Author: ```suggestion
- `redeem(shares, receiver, owner)`: redeems `shares` of the strategy for the underlying asset.
```

- balanceOf(address): return the number of shares of the strategy that the address has
- convertToAssets(shares: uint256): Converts `shares` into the corresponding amount of asset.

> - convertToShares(assets: uint256): Converts `assets` into the corresponding amount of shares.

Author: ```suggestion
- `convertToShares(assets: uint256)`: Converts `assets` into the corresponding amount of shares.
```

> - previewWithdraw(assets: uint256): Converst `assets` into the corresponding amount of shares rounding up.

Author: ```suggestion
- `previewWithdraw(assets: uint256)`: Converst `assets` into the corresponding amount of shares rounding up.
```

This means that the vault can deposit into any ERC4626 vault but also that a non-compliant strategy can be implemented provided that these functions have been implemented (even in a non ERC4626 compliant way). 

## ERC4626 compliance
Vault Shares are ERC4626 compliant. 

The most important implication is that `withdraw` and `redeem` functions as presented in ERC4626, with the ability to add two additional non-standard options.

> 1. max_loss: The amount in basis points that the withdrawer will accept as a loss. I.E. 100 = 1% loss accepted.

Author: ```suggestion
1. `max_loss`: The amount in basis points that the withdrawer will accept as a loss. I.E. 100 = 1% loss accepted.
```

> 2. strategies: This is an array of strategies to use as the withdrawal queue instead of the default queue.

Author: ```suggestion
2. `strategies`: This is an array of strategies to use as the withdrawal queue instead of the default queue.
```

### Shutdown mode
In the case the current roles stop fulfilling their responsibilities or something else happens, the EMERGENCY_MANAGER can shutdown the vault.

The shutdown mode should be the last option in an emergency as it is irreversible. 

### Deposits

> _Light emergency_: Deposits can be paused by setting depositLimit to 0

Author: ```suggestion
`_Light emergency_`: Deposits can be paused by setting depositLimit to 0
```

> _Shutdown mode_: Deposits are not allowed

Author: ```suggestion
`_Shutdown mode_`: Deposits are not allowed
```

### Withdrawals
Withdrawals can't be paused under any circumstance

### Accounting
Shutdown mode does not affect accounting

### Debt rebalance

> _Light emergency_: Setting minimumTotalIdle to MAX_UINT256 will result in the vault requesting the debt back from strategies. This would stop new strategies from getting funded too, as the vault prioritizes minimumTotalIdle

Author: ```suggestion
`_Light emergency_`: Setting minimumTotalIdle to MAX_UINT256 will result in the vault requesting the debt back from strategies. This would stop new strategies from getting funded too, as the vault prioritizes minimumTotalIdle
```

> _Shutdown mode_: All strategies maxDebt is set to 0. Strategies will return funds as soon as they can.

Author: ```suggestion
`_Shutdown mode_`: All strategies maxDebt is set to 0. Strategies will return funds as soon as they can.
```

### Relevant emergency

> In the case the current roles stop fulfilling their responsibilities or something else happens, the EMERGENCY_MANAGER can shutdown the vault. 

Author: ```suggestion
In the case the current roles stop fulfilling their responsibilities or something else happens, the `EMERGENCY_MANAGER` can shutdown the vault. 
```

The shutdown mode should be the last option in an emergency as it is irreversible. 

During shutdown mode, the vault will try to get funds back from every strategy as soon as possible. 

No strategies can be added during shutdown

> Any relevant role will start pointing to the EMERGENCY_MANAGER in case new permissioned allowed actions need to be taken.

Author: ```suggestion
Any relevant role will start pointing to the `EMERGENCY_MANAGER` in case new permissioned allowed actions need to be taken.
```

Deposits are limited by the maxAvailableDeposit function that can be changed by the strategist if non uint256.max values are desired.

#### Withdrawals / Redeems
Users can redeem their shares at any point in time if there is liquidity available. 

The amount of a withdraw or redeem can be limited by the strategist by overriding the maxAvailableWithdraw function.

In order to properly comply with the ERC-4626 standard and still allow losses, both withdraw and redeem have an additional optional parameter of 'maxLoss' that can be used. The default for 'maxLoss' is 0 (i.e. revert if any loss) for withdraws, and 10_000 (100%) for redeems.

#### Strategy Shares
The strategy issues shares to each depositor to track their relative share of assets. Shares are ERC20 transferable yield-bearing tokens.

They are ERC4626 compliant. Please read [ERC4626 compliance](https://hackmd.io/cOFvpyR-SxWArfthhLJb5g#ERC4626-compliance) to understand the implications. 

#### Accounting
The strategy will evaluate profit and losses from the yield generating activities. 

This is done comparing the current totalAssets of the strategy with the amount returned from _harvestAndReport()

> If totalAssets < newTotalAssets: the vault will record a loss

Author: ```suggestion
If `totalAssets` < `newTotalAssets`: the vault will record a loss
```

> If totalAssets > newTotalAssets: the vault will record a profit

Author: ```suggestion
If `totalAssets` > `newTotalAssets`: the vault will record a profit
```
